transaction/src/manifest/lexer.rs

```diff
@@ -1,76 +1,35 @@
+use crate::manifest::diagnostic_snippets::create_snippet;
+use crate::manifest::token::{Position, Span, Token, TokenKind};
+use crate::position;
+// use sbor::rust::fmt;
+use sbor::rust::fmt::Debug;
 use sbor::rust::str::FromStr;
-
-/// The span of tokens. The `start` and `end` are Unicode code points / UTF-32 - as opposed to a
-/// byte-based / UTF-8 index.
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
-pub struct Span {
-    /// The start of the span, exclusive
-    pub start: Position,
-    /// The end of the span, inclusive
-    pub end: Position,
-}
-
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
-pub struct Position {
-    /// A 0-indexed cursor indicating the next unicode char from the start
-    pub full_index: usize,
-    /// A 1-indexed cursor indicating the line number (assuming \n is a line break)
-    pub line_number: usize,
-    /// A 0-indexed cursor indicating the character offset in the line
-    pub line_char_index: usize,
-}
+use std::fmt::Display;
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub enum TokenKind {
-    // ==============
-    // Literals
-    // ==============
-    BoolLiteral(bool),
-    I8Literal(i8),
-    I16Literal(i16),
-    I32Literal(i32),
-    I64Literal(i64),
-    I128Literal(i128),
-    U8Literal(u8),
-    U16Literal(u16),
-    U32Literal(u32),
-    U64Literal(u64),
-    U128Literal(u128),
-    StringLiteral(String),
-
-    Ident(String),
-
-    /* Punctuations */
-    OpenParenthesis,
-    CloseParenthesis,
-    LessThan,
-    GreaterThan,
-    Comma,
-    Semicolon,
-    FatArrow,
+pub enum LexerErrorKind {
+    UnexpectedEof,
+    UnexpectedChar(char),
+    InvalidInteger(String),
+    InvalidUnicode(u32),
 }
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct Token {
-    pub kind: TokenKind,
+pub struct LexerError {
+    pub error_kind: LexerErrorKind,
     pub span: Span,
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub enum LexerError {
-    UnexpectedEof,
-    UnexpectedChar(char, Position),
-    InvalidInteger(String, Position),
-    InvalidUnicode(u32, Position),
-    UnknownIdentifier(String, Position),
-}
-
 #[derive(Debug, Clone)]
 pub struct Lexer {
     /// The input text chars
     text: Vec<char>,
     /// The current position in the text
     current: Position,
+    /// The previous position in the text
+    previous: Position,
+    /// The start position of token being lexed
+    start: Position,
```

Comment:
Although on reflection I don't think we need `start`/`token_start` at all?

I think passing it into the methods that need it will be a little clearer/more specific, and have fewer edge cases/bugs.

Reply:
Got rid of `start`